src/lib/salad/bps_tree.h

```diff
@@ -314,6 +314,14 @@
  * #define BPS_BLOCK_LINEAR_SEARCH
  */
 
+/**
+ * A switch to make the tree store the cardinality of each of its
+ * child blocks in an array. A block cardinality is the amount of
+ * elements in its underlying leaf blocks (or a sum of its child
+ * cardinalities if the children are inner).
+ * #define BPS_BLOCK_CHILD_CARDS_ARRAY
```

Comment:
Abundance of macros makes the code ugly. Let's measure the degradation introduced by this change, both in CPU and memory consumption. If it proves to be insignificant, better enable the feature unconditionally.

Reply:
There're three ways to store cardinalities implemented:
1. Using an Array of Children Cardinalities (AoCC), which was the first approach implemented. We store chlidren cardinalities in `child_cards` arrays of each inner node.
2. Using Inner Block Cardinalities (IBC), which makes the tree store own cardinalities in inner blocks.
3. A hybrid approach using both `child_cards` array and inner block own cardinality, technically the one suggested in the comment below.

So here're benchmarking results.

<details>

<summary>Raw BPS tree benchmarks (perf/bps_tree.cc)...</summary>

| Benchmark               | Base   | AoCC                          | IBC                               | Both                          |
| ----------------------- | ------ | ----------------------------- | --------------------------------- | ----------------------------- |
| find_rand/1M            | 288ns  | 306ns ($\color{red}+6.25％$)   | 290ns ($\color{RedOrange}+0.69％$) | 306ns ($\color{red}+6.25％$)   |
| find_rand/10M           | 528ns  | 562ns ($\color{red}+6.44％$)   | 511ns ($\color{green}-3.22％$)     | 527ns ($\color{green}-0.19％$) |
| find_rand/100M          | 747ns  | 766ns ($\color{red}+2.54％$)   | 748ns ($\color{RedOrange}+0.13％$) | 764ns ($\color{red}+2.28％$)   |
| find_rand/500M          | 978ns  | 1011ns ($\color{red}+3.37％$)  | 982ns ($\color{RedOrange}+0.41％$) | 1023ns ($\color{red}+4.60％$)  |
| delete_insert_rand/1M   | 569ns  | 711ns ($\color{red}+24.96％$)  | 657ns ($\color{red}+15.47％$)      | 704ns ($\color{red}+23.73％$)  |
| delete_insert_rand/10M  | 887ns  | 1017ns ($\color{red}+14.66％$) | 1006ns ($\color{red}+13.42％$)     | 1012ns ($\color{red}+14.09％$) |
| delete_insert_rand/100M | 1176ns | 1360ns ($\color{red}+15.65％$) | 1303ns ($\color{red}+10.80％$)     | 1355ns ($\color{red}+15.22％$) |
| delete_insert_rand/500M | 1439ns | 1686ns ($\color{red}+17.16％$) | 1593ns ($\color{red}+10.70％$)     | 1689ns ($\color{red}+17.37％$) |
| insert_rand/10M         | 490ns  | 537ns ($\color{red}+9.59％$)   | 523ns ($\color{red}+6.73％$)       | 532ns ($\color{red}+8.57％$)   |
| insert_rand/100M        | 771ns  | 854ns ($\color{red}+10.77％$)  | 812ns ($\color{red}+5.32％$)       | 849ns ($\color{red}+10.12％$)  |
| insert_rand/500M        | 1045ns | 1141ns ($\color{red}+9.19％$)  | 1098ns ($\color{red}+5.07％$)      | 1138ns ($\color{red}+8.90％$)  |
| delete_rand/10M         | 555ns  | 610ns ($\color{red}+9.91％$)   | 592ns ($\color{red}+6.67％$)       | 605ns ($\color{red}+9.01％$)   |
| delete_rand/100M        | 819ns  | 898ns ($\color{red}+9.65％$)   | 862ns ($\color{red}+5.25％$)       | 887ns ($\color{red}+8.30％$)   |
| delete_rand/500M        | 1061ns | 1166ns ($\color{red}+9.90％$)  | 1119ns ($\color{red}+5.47％$)      | 1180ns ($\color{red}+11.22％$) |

</details>

<details>

<summary>Measuring RPS using raw sockets, batch size is 1000...</summary>

| Benchmark | Base    | AoCC                                | IBC                             | Both                            |
| --------- | ------- | ----------------------------------- | ------------------------------- | ------------------------------- |
| insert    | 457krps | 440krps ($\color{red}-3.63％$)       | 448krps ($\color{red}-2.02％$)   | 427krps ($\color{red}-6.59％$)   |
| replace   | 451krps | 451krps ($\color{RedOrange}-0.02％$) | 451krps ($\color{green}+0.05％$) | 451krps ($\color{green}+0.16％$) |
| select    | 760krps | 723krps ($\color{red}-4.83％$)       | 737krps ($\color{red}-3.01％$)   | 736krps ($\color{red}-3.22％$)   |
| delete    | 424krps | 415krps ($\color{red}-2.08％$)       | 415krps ($\color{red}-2.00％$)   | 403krps ($\color{red}-4.99％$)   |

</details>

The memory overhead of the IBC solution is negligible. The overhead of the AoCC is a bit more complex, but practically I've got about 2%.

<details>

<summary>Measuring memory consumption (512B block, 16K extent size)</summary>

| Tree configuration | Base       | AoCC                             | IBC                                   |
| ------------------ | ---------- | -------------------------------- | ------------------------------------- |
| 8-byte elem x1M    | 11024KB    | 11264KB ($\color{red}+2.18％$)    | 11040KB ($\color{RedOrange}+0.15％$)   |
| 8-byte elem x10M   | 109904KB   | 112384KB ($\color{red}+2.26％$)   | 110016KB ($\color{RedOrange}+0.10％$)  |
| 8-byte elem x100M  | 1098720KB  | 1123520KB ($\color{red}+2.26％$)  | 1099856KB ($\color{RedOrange}+0.10％$) |
| 8-byte elem x500M  | 5493440KB  | 5617440KB ($\color{red}+2.26％$)  | 5499168KB ($\color{RedOrange}+0.10％$) |
| 16-byte elem x1M   | 22032KB    | 22480KB ($\color{red}+2.03％$)    | 22032KB (-)                           |
| 16-byte elem x10M  | 220048KB   | 224496KB ($\color{red}+2.02％$)   | 220048KB (-)                          |
| 16-byte elem x100M | 2200192KB  | 2244720KB ($\color{red}+2.02％$)  | 2200192KB (-)                         |
| 16-byte elem x500M | 11000768KB | 11223456KB ($\color{red}+2.02％$) | 11000768KB (-)                        |

<details>

I would keep this feature optional but if it'll be decided to apply it to default tree too, it seems the least affecting the current tree option is IBC (only storing own cardinalities in inner blocks).

Reply:
Discussed offline: it's decided to keep both options but use the default tree with no changes for regular tree index to prevent the regressions.